Add support for GitHub Nested Teams 🐣

cblecker opened the pull request:
Add support for Github Nested Teams to peribolos.
https://blog.github.com/2017-06-13-nested-teams-add-depth-to-your-team-structure/

example config:
```
teams:
  sig-foo:
    description: foo enthusiasts
    members:
    - george
    - jane
    privacy: closed
    teams:
      sig-foo-leads:
      members:
      - george
      sig-foo-pr-reviews:
      members:
      - jane
```

fejta commented on prow/cmd/peribolos/main.go:
```diff
@@ -219,36 +219,46 @@ func dumpOrgConfig(client dumpClient, orgName string) (*org.Config, error) {
-		}
-		for _, m := range teamMembers {
-			nt.Members = append(nt.Members, m.Login)
+	out.Teams = make(map[string]org.Team)
```
I think the general goal here is:

a) Convert github.Team metadata into org.Team metadata
b) Figure out what children a given parent has
c) Create the tree of children, starting at ancestors.

So then we wind up with something like this:
```
names := map[int]string{} // what's the name of a team?
idMap := map[int]org.Team{} // metadata for a team
children := map[int][]int // what children does it have
var tops []int // what are the top-level teams

// first create all the org.Team items
for _, t := range teams {
  nt := org.Team{
      TeamMetadata: org.TeamMetadata{
         Description: t.Description
      },
      Maintainer: client.ListTeamMembers(t.ID)
  }
  names[t.ID] = t.Name
  idMap[t.ID] = nt
  if t.ParentID == 0 { // top level team
    tops = append(tops, t.ID)
  } else { // add this id to the list of the parent's children
    children[t.ParentID] = append(children[t.ParentID], t.ID)
  }
}

makeChild := func(id int) org.Team {
  t := idMap[id]
  for _, cid := range children[id] {
    child := makeChild(cid)
    t.Children = append(t.Children, child)
  }
  return t
} 

for _, id := range tops {
  out.Teams[names[id]] = makeChild(id)
}
```

cblecker commented:
@fejta I think this is ready for another review. Got the dump working (thanks for the help!). Example from kubernetes-sigs:
```
  test-parent-team:
    description: ""
    maintainers:
    - cblecker
    privacy: closed
    teams:
      test-child-team:
        description: ""
        maintainers:
        - cblecker
        privacy: closed
```

fejta commented on prow/github/types.go:
```diff
@@ -575,10 +575,12 @@ type Content struct {
+	Name         string `json:"name,omitempty"`
+	Description  string `json:"description,omitempty"`
+	Privacy      string `json:"privacy,omitempty"`
+	Parent       *Team  `json:"parent,omitempty"`
```
Maybe note with a comment that this is only relevant for responses

cblecker replied:
Done.

fejta commented on prow/cmd/peribolos/main_test.go:
```diff
@@ -1077,7 +1077,7 @@ func TestConfigureTeam(t *testing.T) {
 		t.Run(tc.name, func(t *testing.T) {
 			fc := makeFakeTeamClient(tc.github)
-			err := configureTeam(fc, fakeOrg, tc.teamName, tc.config, tc.github)
+			err := configureTeam(fc, fakeOrg, tc.teamName, tc.config, tc.github, 0)
```
Will you `s/0/tc.parentID` and ensure cover this in "do not patch team when values are the same" and have a "patch team when parent changes"

fejta commented on prow/cmd/peribolos/main.go:
```diff
@@ -675,6 +721,11 @@ func configureTeam(client editTeamClient, orgName, teamName string, team org.Tea
 	} else {
 		gt.Privacy = ""
 	}
+	if parent > 0 && gt.ParentTeamID != &parent {
```
Can I make a child team into a top-level team by later setting its parentID to 0? I think so, in which case we probably don't need the `parent > 0` check?

cblecker replied:
no.. you actually need to send `parent_team_id: null` for that to work. removed the omitempty so it will handle this case.

cblecker commented:
@fejta Addressed comments, but having some issues with the tests.. would you mind having a look at 9a6de59a713f6ab2de952d2fe7be294fed7b8c2b ? there might be something dumb and obvious I'm missing.. both the test cases you call out are failing:
```
--- FAIL: TestConfigureTeam (0.00s)
    --- FAIL: TestConfigureTeam/patch_team_when_parent_changes (0.00s)
        main_test.go:1130: actual {3 whatever   0xc420129680 <nil>} != expected {3 whatever   <nil> 0xc42023b0b0}
    --- FAIL: TestConfigureTeam/do_not_patch_team_when_values_are_the_same (0.00s)
        main_test.go:1125: unexpected error: failed to edit random-org team 4(fail): injected name failure
```

fejta commented on prow/cmd/peribolos/main.go:
```diff
@@ -655,7 +701,7 @@ type editTeamClient interface {
 
 // configureTeam patches the team name/description/privacy when values differ
-func configureTeam(client editTeamClient, orgName, teamName string, team org.Team, gt github.Team) error {
+func configureTeam(client editTeamClient, orgName, teamName string, team org.Team, gt github.Team, parent *int) error {
```
I do not understand the motivation behind / value of changing this to `parent *int` instead of `int`.

cblecker replied:
`0` is technically a valid parent team ID. We need to be able to tell the difference between `0` and `nil`.

fejta replied:
got it, thanks! Can you include a unit test for removing the parent from a team then?

fejta commented on prow/cmd/peribolos/main_test.go:
```diff
@@ -698,6 +698,13 @@ func (c *fakeTeamClient) EditTeam(team github.Team) (*github.Team, error) {
 		t.Privacy = team.Privacy
 	}
+	if team.ParentTeamID != nil {
+		team.Parent = &github.Team{
```
Wrong lvalue

cblecker replied:
I'm confused.. I don't think it is. The goal here in the fake client is to take ParentTeamID, and set the Parent struct to contain a fake team with that matching ID.

fejta replied:
Agreed, and what happens to `team.Parent` when this method returns? Compare the lvalue in this `if` statement to the others.

fejta commented on prow/cmd/peribolos/main_test.go:
```diff
@@ -974,17 +981,20 @@ func TestConfigureTeam(t *testing.T) {
 		name     string
 		err      bool
 		teamName string
+		parent   *int
```
where do you use this value?

fejta commented on prow/cmd/peribolos/main_test.go:
```diff
@@ -1035,9 +1047,29 @@ func TestConfigureTeam(t *testing.T) {
+			config:   org.Team{},
+			github: github.Team{
+				ID:   3,
+				Name: whatev,
```
Please make sure the parent ID is the only bad value.

By making `config.Name != github.Name` also differ it is no longer clear that `ParentTeamID` change causes an update (might just be luck due to the name diff).

fejta commented on prow/cmd/peribolos/main_test.go:
```diff
@@ -1035,9 +1047,29 @@ func TestConfigureTeam(t *testing.T) {
+				Name: whatev,
+				Parent: &github.Team{
+					ID:   4,
+					Name: "sup",
```
Technically `Name` is unnecessary here. Consider removing for clarity

cblecker commented:
@fejta oooookay. I think I got this figured out, and the tests working. PTAL, and I'll squash after.. thanks again for the pointers! :)

fejta commented:
commit includes an emoji, auto
/approve

cblecker commented:
Squashified! :shipit:

BenTheElder commented:
/test pull-test-infra-bazel-canary

BenTheElder commented:
Sorry for the noise, this is the failure I'm looking for, after
https://github.com/kubernetes/test-infra/pull/8525 hopefully we'll have the
fix confirmed.

On Sat, Jun 30, 2018 at 11:34 AM k8s-ci-robot <notifications@github.com>
wrote:

> @cblecker <https://github.com/cblecker>: The following test *failed*, say
> /retest to rerun them all:
> Test name Commit Details Rerun command
> pull-test-infra-bazel-canary 7f86b7a
> <https://github.com/kubernetes/test-infra/commit/7f86b7a837c594320094db510606dc977a177f97>
> link
> <https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/test-infra/8486/pull-test-infra-bazel-canary/197/> /test
> pull-test-infra-bazel-canary
>
> Full PR test history
> <https://k8s-gubernator.appspot.com/pr/test-infra/8486>. Your PR dashboard
> <https://k8s-gubernator.appspot.com/pr/cblecker>. Please help us cut down
> on flakes by linking to
> <https://git.k8s.io/community/contributors/devel/flaky-tests.md#filing-issues-for-flaky-tests>
> an open issue
> <https://github.com/kubernetes/test-infra/issues?q=is:issue+is:open> when
> you hit one in your PR.
>
> Instructions for interacting with me using PR comments are available here
> <https://git.k8s.io/community/contributors/guide/pull-requests.md>. If
> you have questions or suggestions related to my behavior, please file an
> issue against the kubernetes/test-infra
> <https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:>
> repository. I understand the commands that are listed here
> <https://go.k8s.io/bot-commands>.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/test-infra/pull/8486#issuecomment-401558581>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA4Bqx3OSeCW6Qy4QSqfM3e1DH0Uh3_Yks5uB8SlgaJpZM4U6cQB>
> .
>

cblecker commented on prow/cmd/peribolos/main.go:
```diff
@@ -245,8 +250,32 @@ func dumpOrgConfig(client dumpClient, orgName string) (*org.Config, error) {
+		return t
+	}
+
+	out.Teams = make(map[string]org.Team, len(tops))
```
One last change.. moved this line down here from L222

BenTheElder commented:
/test pull-test-infra-bazel-canary

cblecker commented:
/hold cancel

fejta commented:
/retest
